File: dings/src/messaging/message_generator.py
"""LLM-based message generation with constraint enforcement."""

from __future__ import annotations

import hashlib
import logging
from typing import Protocol

from .config import MAX_GENERATION_RETRIES
from .listing_analyzer import ListingAnalyzer
from .models import (
    FollowUpStage,
    ListingSignals,
    Message,
    MessageVariant,
)
from .personalization import PersonalizationEngine
from .spam_guard import SpamGuard
from .templates import build_followup_prompt, build_generation_prompt

logger = logging.getLogger(__name__)


class LLMClient(Protocol):
    """Protocol for LLM clients used by the generator."""

    def generate(self, system_prompt: str, user_prompt: str) -> str: ...


class MessageGenerator:
    """Generates personalized messages for Kleinanzeigen listings.

    Full pipeline: analyze listing → personalize → generate via LLM → validate.
    """

    def __init__(self, llm_client: LLMClient):
        self._llm = llm_client
        self._analyzer = ListingAnalyzer()
        self._personalizer = PersonalizationEngine()
        self._spam_guard = SpamGuard(llm_client=llm_client)
        self._sent_hashes: set[str] = set()

    def generate(
        self,
        raw_listing_text: str,
        listing_id: str = "",
        listing_url: str = "",
        variant: MessageVariant | None = None,
    ) -> Message:
        """Generate a personalized initial message for a listing.

        Args:
            raw_listing_text: The full text content of the listing.
            listing_id: Optional ID for tracking.
            listing_url: Optional URL for tracking.
            variant: Force a specific variant, or None to auto-select best fit.

        Returns:
            A validated Message ready to send.

        Raises:
            MessageGenerationError: If generation fails after all retries.
        """
        # Step 1: Analyze listing
        signals = self._analyzer.analyze(raw_listing_text, listing_id, listing_url)

        # Step 2: Personalize
        personalization = self._personalizer.personalize(signals)

        # Step 3: Select variant
        if variant is None:
            variant = personalization.recommended_variants[0]

        # Step 4: Generate + validate (with retries)
        return self._generate_with_retries(signals, personalization, variant)

    def generate_all_variants(
        self,
        raw_listing_text: str,
        listing_id: str = "",
        listing_url: str = "",
    ) -> dict[MessageVariant, Message]:
        """Generate messages for all 6 variants (useful for review/testing).

        Returns a dict of variant -> Message. Variants that fail validation
        after all retries are excluded.
        """
        signals = self._analyzer.analyze(raw_listing_text, listing_id, listing_url)
        personalization = self._personalizer.personalize(signals)

        results: dict[MessageVariant, Message] = {}
        for v in MessageVariant:
            try:
                msg = self._generate_with_retries(signals, personalization, v)
                results[v] = msg
            except MessageGenerationError:
                logger.warning("Variant %s failed all retries for listing %s", v.name, listing_id)

        return results

    def generate_followup(
        self,
        raw_listing_text: str,
        stage: FollowUpStage,
        listing_id: str = "",
        listing_url: str = "",
    ) -> Message:
        """Generate a follow-up message.

        Args:
            raw_listing_text: The full listing text (same as initial).
            stage: FOLLOWUP_1 or FOLLOWUP_2.
            listing_id: For tracking.
            listing_url: For tracking.

        Returns:
            A validated follow-up Message.
        """
        if stage not in (FollowUpStage.FOLLOWUP_1, FollowUpStage.FOLLOWUP_2):
            raise ValueError(f"Invalid follow-up stage: {stage}")

        signals = self._analyzer.analyze(raw_listing_text, listing_id, listing_url)
        stage_num = 1 if stage == FollowUpStage.FOLLOWUP_1 else 2

        system_prompt, user_prompt = build_followup_prompt(signals, stage_num)

        for attempt in range(1, MAX_GENERATION_RETRIES + 2):
            raw_message = self._llm.generate(system_prompt, user_prompt).strip()
            raw_message = self._clean_message(raw_message)

            # Follow-ups have relaxed validation (no personalization check needed)
            validation = self._spam_guard.validate(raw_message, signals)

            if validation.passed and not self._is_duplicate(raw_message):
                self._sent_hashes.add(self._hash(raw_message))
                return Message(
                    text=raw_message,
                    variant=MessageVariant.SPECIFIC_OBSERVER,  # Not relevant for follow-ups
                    listing_id=listing_id,
                    listing_url=listing_url,
                    spam_guard_score=validation.score,
                    generation_attempt=attempt,
                    stage=stage,
                )

            if attempt <= MAX_GENERATION_RETRIES:
                # Add rejection feedback to prompt for next attempt
                user_prompt = (
                    user_prompt
                    + f"\n\nVORHERIGER VERSUCH ABGELEHNT: {validation.rejection_reason}"
                    + "\nBitte korrigiere diese Probleme."
                )
                logger.info(
                    "Follow-up attempt %d rejected: %s",
                    attempt,
                    validation.rejection_reason,
                )

        raise MessageGenerationError(
            f"Follow-up generation failed after {MAX_GENERATION_RETRIES + 1} attempts"
        )

    def analyze_listing(self, raw_listing_text: str) -> ListingSignals:
        """Expose the analyzer for external use (e.g., debugging)."""
        return self._analyzer.analyze(raw_listing_text)

    def _generate_with_retries(
        self,
        signals: ListingSignals,
        personalization,
        variant: MessageVariant,
    ) -> Message:
        system_prompt, user_prompt = build_generation_prompt(
            signals, personalization, variant
        )

        for attempt in range(1, MAX_GENERATION_RETRIES + 2):
            raw_message = self._llm.generate(system_prompt, user_prompt).strip()
            raw_message = self._clean_message(raw_message)

            validation = self._spam_guard.validate(raw_message, signals)

            if validation.passed and not self._is_duplicate(raw_message):
                self._sent_hashes.add(self._hash(raw_message))
                return Message(
                    text=raw_message,
                    variant=variant,
                    listing_id=signals.listing_id,
                    listing_url=signals.listing_url,
                    spam_guard_score=validation.score,
                    generation_attempt=attempt,
                )

            if attempt <= MAX_GENERATION_RETRIES:
                user_prompt = (
                    user_prompt
                    + f"\n\nVORHERIGER VERSUCH ABGELEHNT: {validation.rejection_reason}"
                    + "\nBitte korrigiere diese Probleme."
                )
                logger.info(
                    "Variant %s attempt %d rejected: %s",
                    variant.name,
                    attempt,
                    validation.rejection_reason,
                )

        raise MessageGenerationError(
            f"Variant {variant.name} failed after {MAX_GENERATION_RETRIES + 1} attempts "
            f"for listing {signals.listing_id}"
        )

    def _clean_message(self, text: str) -> str:
        """Remove LLM artifacts like quotes, prefixes, etc."""
        # Remove surrounding quotes
        if text.startswith('"') and text.endswith('"'):
            text = text[1:-1]
        if text.startswith("'") and text.endswith("'"):
            text = text[1:-1]
        # Remove common LLM preambles
        prefixes_to_strip = [
            "Hier ist die Nachricht:",
            "Nachricht:",
            "Hier ist mein Vorschlag:",
        ]
        for prefix in prefixes_to_strip:
            if text.startswith(prefix):
                text = text[len(prefix):].strip()
        return text.strip()

    def _hash(self, message: str) -> str:
        normalized = " ".join(message.lower().split())
        return hashlib.sha256(normalized.encode()).hexdigest()[:16]

    def _is_duplicate(self, message: str) -> bool:
        return self._hash(message) in self._sent_hashes


class MessageGenerationError(Exception):
    """Raised when message generation fails after all retries."""

File: packages/llm/src/messaging/message-generator.ts
import { createHash } from "crypto";
import { MAX_GENERATION_RETRIES } from "./config";
import { DelayCalculator } from "./delay-calculator";
import { ListingAnalyzer } from "./listing-analyzer";
import {
	type ListingSignals,
	type Message,
	FollowUpStage,
	MessageVariant,
	createMessage,
} from "./models";
import { PersonalizationEngine } from "./personalization";
import { PostProcessor } from "./post-processor";
import { Safeguard } from "./safeguard";
import { SpamGuard } from "./spam-guard";
import { buildFollowupPrompt, buildGenerationPrompt } from "./templates";

const SKIP_TOKEN = "[SKIP]";

export interface LLMClient {
	generate(systemPrompt: string, userPrompt: string): Promise<string>;
}

export interface MessageResult {
	message: Message | null;
	skipped: boolean;
	delayMs: number;
}

export class MessageGenerationError extends Error {}

export class MessageGenerator {
	private llm: LLMClient;
	private analyzer = new ListingAnalyzer();
	private personalizer = new PersonalizationEngine();
	private spamGuard: SpamGuard;
	private postProcessor = new PostProcessor();
	private safeguard: Safeguard;
	private delay: DelayCalculator;
	private sentHashes = new Set<string>();

	constructor(llmClient: LLMClient, opts: { testMode?: boolean } = {}) {
		this.llm = llmClient;
		this.spamGuard = new SpamGuard(llmClient);
		this.safeguard = new Safeguard(llmClient);
		this.delay = new DelayCalculator({ testMode: opts.testMode });
	}

	async generate(
		rawListingText: string,
		listingId = "",
		listingUrl = "",
		variant?: MessageVariant,
	): Promise<MessageResult> {
		const signals = this.analyzer.analyze(
			rawListingText,
			listingId,
			listingUrl,
		);
		const personalization = this.personalizer.personalize(signals);

		if (!variant) {
			variant = personalization.recommendedVariants[0];
		}

		return this.generateWithRetries(signals, personalization, variant, true);
	}

	async generateAllVariants(
		rawListingText: string,
		listingId = "",
		listingUrl = "",
	): Promise<Map<MessageVariant, MessageResult>> {
		const signals = this.analyzer.analyze(
			rawListingText,
			listingId,
			listingUrl,
		);
		const personalization = this.personalizer.personalize(signals);

		const results = new Map<MessageVariant, MessageResult>();
		for (const v of Object.values(MessageVariant)) {
			try {
				const result = await this.generateWithRetries(
					signals,
					personalization,
					v as MessageVariant,
					true,
				);
				results.set(v as MessageVariant, result);
			} catch (e) {
				if (e instanceof MessageGenerationError) {
					console.warn(
						`Variant ${v} failed all retries for listing ${listingId}`,
					);
				} else {
					throw e;
				}
			}
		}
		return results;
	}

	async generateFollowup(
		rawListingText: string,
		stage: FollowUpStage,
		listingId = "",
		listingUrl = "",
	): Promise<MessageResult> {
		if (
			stage !== FollowUpStage.FOLLOWUP_1 &&
			stage !== FollowUpStage.FOLLOWUP_2
		) {
			throw new Error(`Invalid follow-up stage: ${stage}`);
		}

		const signals = this.analyzer.analyze(
			rawListingText,
			listingId,
			listingUrl,
		);
		const stageNum = stage === FollowUpStage.FOLLOWUP_1 ? 1 : 2;

		let [systemPrompt, userPrompt] = buildFollowupPrompt(signals, stageNum);

		for (let attempt = 1; attempt <= MAX_GENERATION_RETRIES + 1; attempt++) {
			const rawMessage = this.cleanMessage(
				(await this.llm.generate(systemPrompt, userPrompt)).trim(),
			);

			if (this.isSkip(rawMessage)) {
				console.info(`Follow-up for ${listingId} skipped by agent decision`);
				const delayResult = this.delay.calculate(0, false);
				return { message: null, skipped: true, delayMs: delayResult.delayMs };
			}

			// Pipeline: PostProcess → SpamGuard → Safeguard
			const processed = this.postProcessor.process(rawMessage);
			const validation = await this.spamGuard.validate(processed, signals);

			if (validation.passed && !this.isDuplicate(processed)) {
				const safeguardResult = await this.safeguard.check(processed);
				if (!safeguardResult.passed) {
					if (attempt <= MAX_GENERATION_RETRIES) {
						userPrompt +=
							`\n\nVORHERIGER VERSUCH ABGELEHNT: ${safeguardResult.reason}` +
							"\nBitte korrigiere diese Probleme.";
						console.info(
							`Follow-up attempt ${attempt} rejected by safeguard: ${safeguardResult.reason}`,
						);
						continue;
					}
				}

				this.sentHashes.add(this.hash(processed));
				this.delay.markActive();
				const delayResult = this.delay.calculate(processed.length, false);

				const message = createMessage(
					processed,
					MessageVariant.DIRECT_HONEST,
					listingId,
					{
						listingUrl,
						spamGuardScore: validation.score,
						generationAttempt: attempt,
						stage,
					},
				);
				return { message, skipped: false, delayMs: delayResult.delayMs };
			}

			if (attempt <= MAX_GENERATION_RETRIES) {
				userPrompt +=
					`\n\nVORHERIGER VERSUCH ABGELEHNT: ${validation.rejectionReasons.join("; ")}` +
					"\nBitte korrigiere diese Probleme.";
				console.info(
					`Follow-up attempt ${attempt} rejected: ${validation.rejectionReasons.join("; ")}`,
				);
			}
		}

		throw new MessageGenerationError(
			`Follow-up generation failed after ${MAX_GENERATION_RETRIES + 1} attempts`,
		);
	}

	analyzeListing(rawListingText: string): ListingSignals {
		return this.analyzer.analyze(rawListingText);
	}

	private isSkip(message: string): boolean {
		const cleaned = message.trim().toUpperCase();
		return cleaned === SKIP_TOKEN || cleaned === "[SKIP]";
	}

	private async generateWithRetries(
		signals: ListingSignals,
		personalization: ReturnType<PersonalizationEngine["personalize"]>,
		variant: MessageVariant,
		isFirstInConversation: boolean,
	): Promise<MessageResult> {
		let [systemPrompt, userPrompt] = buildGenerationPrompt(
			signals,
			personalization,
			variant,
		);

		for (let attempt = 1; attempt <= MAX_GENERATION_RETRIES + 1; attempt++) {
			const rawMessage = this.cleanMessage(
				(await this.llm.generate(systemPrompt, userPrompt)).trim(),
			);

			if (this.isSkip(rawMessage)) {
				console.info(
					`Variant ${variant} for ${signals.listingId} skipped by agent decision`,
				);
				const delayResult = this.delay.calculate(0, isFirstInConversation);
				return { message: null, skipped: true, delayMs: delayResult.delayMs };
			}

			// Pipeline: PostProcess → SpamGuard → Safeguard
			const processed = this.postProcessor.process(rawMessage);
			const validation = await this.spamGuard.validate(processed, signals);

			if (validation.passed && !this.isDuplicate(processed)) {
				const safeguardResult = await this.safeguard.check(processed);
				if (!safeguardResult.passed) {
					if (attempt <= MAX_GENERATION_RETRIES) {
						userPrompt +=
							`\n\nVORHERIGER VERSUCH ABGELEHNT: ${safeguardResult.reason}` +
							"\nBitte korrigiere diese Probleme.";
						console.info(
							`Variant ${variant} attempt ${attempt} rejected by safeguard: ${safeguardResult.reason}`,
						);
						continue;
					}
				}

				this.sentHashes.add(this.hash(processed));
				this.delay.markActive();
				const delayResult = this.delay.calculate(
					processed.length,
					isFirstInConversation,
				);

				const message = createMessage(processed, variant, signals.listingId, {
					listingUrl: signals.listingUrl,
					spamGuardScore: validation.score,
					generationAttempt: attempt,
				});
				return { message, skipped: false, delayMs: delayResult.delayMs };
			}

			if (attempt <= MAX_GENERATION_RETRIES) {
				userPrompt +=
					`\n\nVORHERIGER VERSUCH ABGELEHNT: ${validation.rejectionReasons.join("; ")}` +
					"\nBitte korrigiere diese Probleme.";
				console.info(
					`Variant ${variant} attempt ${attempt} rejected: ${validation.rejectionReasons.join("; ")}`,
				);
			}
		}

		throw new MessageGenerationError(
			`Variant ${variant} failed after ${MAX_GENERATION_RETRIES + 1} attempts for listing ${signals.listingId}`,
		);
	}

	private cleanMessage(text: string): string {
		let cleaned = text;
		if (cleaned.startsWith('"') && cleaned.endsWith('"'))
			cleaned = cleaned.slice(1, -1);
		if (cleaned.startsWith("'") && cleaned.endsWith("'"))
			cleaned = cleaned.slice(1, -1);

		const prefixes = [
			"Hier ist die Nachricht:",
			"Nachricht:",
			"Hier ist mein Vorschlag:",
		];
		for (const prefix of prefixes) {
			if (cleaned.startsWith(prefix)) {
				cleaned = cleaned.slice(prefix.length).trim();
			}
		}
		return cleaned.trim();
	}

	private hash(message: string): string {
		const normalized = message.toLowerCase().replace(/\s+/g, " ").trim();
		return createHash("sha256").update(normalized).digest("hex").slice(0, 16);
	}

	private isDuplicate(message: string): boolean {
		return this.sentHashes.has(this.hash(message));
	}
}
